### PR TITLE
execApi -> useApi

### PR DESCRIPTION
**Pull request by krishnaglick**
## Change description

This should be functionally transparent, but all references to execApi have been replaced with either a new Client for server-side calls, or the client object stored in a react context.

There's some minor cleanup work I did while going through everything, but I tried to keep those to a minimum.

### Development

- [X] Application changes have been tested appropriately

### Impact

- [X] Code follows company security practices and guidelines
- [X] Security impact of change has been considered
- [X] Performance impact of change has been considered
- [X] Possible migration needs considered (model migrations, config migrations, etc.)

Please explain any security, performance, migration, or other impacts if relevant:

### Code review

- [X] Pull request has a descriptive title and context useful to a reviewer. Screenshots or screencasts are attached where applicable.
- [X] Relevant tags have been added to the PR (bug, enhancement, internal, etc.)

### CODE REVIEW

File: ui/ui-components/client/client.ts [hunk 1]
```diff
@@ -107,33 +112,39 @@ export class Client {
...
       return window.localStorage.getItem("session:csrfToken");
     }
-  }
+  };
+
+  private static readonly optionDefaults = {
+    useCache: true,
+    errorHandler,
+    uploadHandler,
+  };
```
Comment (edmundito) [inline ui/ui-components/client/client.ts]: Should probably go at the top of the class?

File: ui/ui-components/client/client.ts [hunk 2]
```diff
@@ -107,33 +112,39 @@ export class Client {
...
+  public request = async <Response = any>(
     verb: Method = "get",
     path: string,
     data: AxiosRequestConfig["data"] = {},
-    useCache = true,
-    uploadHandler?: UploadHandler,
-    req?: IncomingMessage,
-    res?: ServerResponse
-  ): Promise<Response> {
+    options: Partial<typeof Client.optionDefaults> = {}
```
Comment (edmundito) [inline ui/ui-components/client/client.ts]: Maybe the Options should be explicit instead of derived from `optionDefaults`. There is no optionDefaults can be broken at any time without having the right Options contract

File: ui/ui-components/client/client.ts [hunk 3]
```diff
@@ -147,7 +158,7 @@ export class Client {
         const uploadPercentage = Math.round(
           (progressEvent.loaded / progressEvent.total) * 100
         );
-        return uploadHandler ? uploadHandler.set({ uploadPercentage }) : null;
+        return options.uploadHandler.set({ uploadPercentage });
```
Comment (edmundito) [inline ui/ui-components/client/client.ts]: I put a note on the story that we can get rid of the uploadHandler entirely because we no longer upload files. It would also be fine to save it for another change.

File: ui/ui-components/client/client.ts [hunk 4]
```diff
@@ -210,16 +221,32 @@ export class Client {
...
     }
-  }
+  };
+}
+
+export const generateClient = (
+  ctx: AppContext | NextPageContext | GetServerSidePropsContext
+) => new Client(getRequestContext(ctx));
+
+function isAppContext(ctx: unknown): ctx is AppContext {
```
Comment (edmundito) [inline ui/ui-components/client/client.ts]: This function has nothing to do with the client, should probably be a util

File: ui/ui-components/contexts/api.tsx [hunk 5]
```diff
@@ -0,0 +1,5 @@
+import { createContext, useContext } from "react";
+import type { Client } from "../client/client";
+
+export const ApiContext = createContext<{ client: Client }>(undefined);
+export const useApi = () => useContext(ApiContext);
```
Comment (edmundito) [inline ui/ui-components/contexts/api.tsx:7]: After looking at the code. I'm torn if this particular function should be in `/hooks/useApi` along the rest of the other context functions. The implementation of it as a context seems irrelevant when it's being used.

File: ui/ui-components/pages/_app.tsx [hunk 6]
```diff
@@ -47,8 +44,16 @@ export default function GrouparooNextApp(
...
   }, [props]);
 
+  const client = typeof window === "undefined" ? props.client : new Client();
   return renderNestedContextProviders(
-    [[WebAppContext, pageContext]],
+    [
+      [WebAppContext, pageContext],
+      [
+        ApiContext,
+        // The serialized object passed from the server is not an isntance of the Client class
```
Comment (edmundito) [inline ui/ui-components/pages/_app.tsx]: This comment could probably be moved above the client declaration

File: ui/ui-components/pages/_app.tsx [hunk 7]
```diff
@@ -108,10 +114,12 @@ GrouparooNextApp.getInitialProps = async (appContext: AppContext) => {
       navigation: navigationResponse.navigation,
       clusterName: navigationResponse.clusterName,
       hydrationError,
+      client,
     };
   } catch (_error) {
     const { formattedErrorMessage, formattedErrorObject } = renderError(_error);
     return {
+      client,
```
Comment (edmundito) [inline ui/ui-components/pages/_app.tsx:119]: To confirm: This client will only be used the first time the page renders on the server side?

File: ui/ui-components/client/client.ts [hunk 8]
```diff
@@ -210,16 +221,32 @@ export class Client {
...
       } else {
+        options.errorHandler.set({ message: error });
         throw error;
       }
     }
-  }
+  };
+}
+
+export const generateClient = (
```
Comment (edmundito) [inline ui/ui-components/client/client.ts:241]: Could also be `getClient`

File: ui/ui-components/client/client.ts (hunk 2 above)
Comment (krishnaglick) [reply ui/ui-components/client/client.ts]: I'm not sure what you're getting at here. optionDefaults are defaulted, and you can't pass anything not in it. They're override only.

File: ui/ui-components/contexts/api.tsx (hunk 5 above)
Comment (krishnaglick) [reply ui/ui-components/contexts/api.tsx:7]: This could potentially just go into `WebAppContext` but I'd have to find-replace both the useApi and the import and that would be a _lot_ of gruntwork for not much gain. This is a context, we should keep it here.

File: ui/ui-components/pages/_app.tsx (hunk 7 above)
Comment (krishnaglick) [reply ui/ui-components/pages/_app.tsx:119]: I'll be pushing up a slight logic change. As long as this client is a client and not some serialized object, it will be used.

File: ui/ui-components/client/client.ts (hunk 2 above)
Comment (edmundito) [reply ui/ui-components/client/client.ts]: My point is that it's strange that this is deriving which options are valid from a constant that assigns default values Using explicit typing is a lot stronger way to establish what's valid.

File: ui/ui-components/pages/_app.tsx (hunk 7 above)
Comment (edmundito) [reply ui/ui-components/pages/_app.tsx:119]: It will always get a serialized object when the browser side runs. I'm wondering if we ever need to share this instance with the page server side.

File: ui/ui-components/client/client.ts (hunk 2 above)
Comment (krishnaglick) [reply ui/ui-components/client/client.ts]: Fair enough, added an interface.

File: ui/ui-components/pages/_app.tsx (hunk 7 above)
Comment (krishnaglick) [reply ui/ui-components/pages/_app.tsx:119]: This is run on the client-side as well.
We should perhaps migrate to [getServerSideProps](https://nextjs.org/docs/basic-features/data-fetching/get-server-side-props) instead.

**Comment by evantahler** [comment]
@krishnaglick can you call out in a self-review what some of the new improvements are?  I see how's much easier to use in components now (eg not top-level pages), but I'm looking for new/better error handling and missing it :/

**Review by edmundito — APPROVED** [review]

**Comment by krishnaglick** [comment]
> @krishnaglick can you call out in a self-review what some of the new improvements are? I see how's much easier to use in components now (eg not top-level pages), but I'm looking for new/better error handling and missing it :/

The biggest change to the error handling was partially from Edmundo's work to centralize the errorHandler. In addition, the client.request call allows you to override the errorHandler in the options if you want or need to.
The goal of this PR was to remove execApi since it _looked_ like a React hook but actually wasn't. This turns it into a real hook and should make its functionality far more expected for engineers, new or otherwise.

**Review by evantahler — APPROVED** [review]
let's give this a try!
